design/specs/specs-null-values.md

```diff
@@ -0,0 +1,170 @@
+---
+title: NULL Value Specs
+description: 
+published: true
+date: 2021-11-25T09:00:00.201Z
+tags: 
+editor: markdown
+dateCreated: 2021-11-25T09:00:00.201Z
+---
+
+## Context
+
+A `NULL` value indicates that a value does not exist in a database. It is important to clarify that `NULL` is not the same as blank or zero. Ideally, we want users to think of `NULL` values as placeholders for unknown or missing values applicable otherwise.
+
+## Prototype
+
+### Link
+
+[Prototype](https://mathesar-prototype.netlify.app/?activeTable=2)
+
+### Videos
+
+- Scenario 1  [Video](https://www.loom.com/share/b12f7634961e4cf29e258d2e46f2a7be)
+- Scenario 2: [Video](https://www.loom.com/share/bd127b3214794cd5bf22a146a1ea7ae8)
+- Scenario 3: [Video](https://www.loom.com/share/c998551736104417ac671b5d556f3365)
+
+## User Experience
+
+### Considerations for handling `NULL` Values
+
+When presented in tables, `NULL` values should be visually distinct so that the user understands that they are not the same as other string types. The style should attempt to communicate the absence of a value, similar to how placeholders are used in text input fields.
+As for the process of transforming values to and from `NULL`, it will be important to have a consistent way of doing this across different data types.
+
+## Scenarios
+
+### Scenario 1: User identifies a `NULL` Value
+
+#### Scenario 1a: The user identifies a `NULL` value in a table
+
+##### Steps for 1a
+
+1. The user opens a table that contains `NULL` values.
+2. The user identifies a cell that contains a `NULL` value.
+    - The cell contains the word `NULL` in all uppercase letters
+    - The cell has a lighter background-color
+    - The cell has a muted text color
+    - The cell has a different font style (italic)
+
+#### Scenario 1b: The user identifies a `NULL` value in a record preview
+
+##### Steps for 1b
+
+1. The user edits a value in a column that links to records from another table.
+2. The user sees the list of records displayed in the record selection dropdown.
+3. The user identifies a record that contains a `NULL` value.
+    - The field contains the word `NULL` in all uppercase letters
+    - The text has a muted text color
+    - The text has a different font style (italic)
+
+#### Scenario 1c: The user identifies a `NULL` value in a boolean type column in checkbox display mode
+
+##### Steps for 1c
+
+1. The user opens a table that contains `NULL` values.
+2. The user identifies a cell that contains an indeterminate state checkbox.
+3. The user hovers over the indeterminate checkbox, and a tooltip with the legend `NULL` is displayed.
+
+### Scenario 2: User edits a `NULL` Value
+
+#### Scenario 2a: The data type is text
+
+##### Steps for 2a
+
+1. The user edits a `NULL` value in a column set to text data type.
+2. The user double-clicks the value cell to enter edit mode.
```

Comment:
I'm confused about this - I thought we had consensus in the comment thread that we should only require a single click to enter edit mode?

See these three comments: 
- https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r765376340
- https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r765211660
- https://github.com/centerofci/mathesar-wiki/pull/13#discussion_r770396298

Reply:
@kgodey When you click on a cell once, the cell gets selected (ie) goes into the select mode, it would be a text box awaiting a text input rather than a greyed-out box without a cursor. But for convenience, when the user starts typing(any keystroke other than delete), the cell will change the mode, so it enters into edit mode with the cell turning into a text box, similar to what would have happened if you had double-clicked the cell

Reply:
@silentninja That makes sense to me, but then this part (later in the spec) becomes confusing to me:

> While in `Select Mode`, pressing the `delete` or `backspace` key will delete the cell state and convert it to a `NULL` value.

I would expect that if typing converts it to edit mode, delete/backspace will remove a single character, not convert it to `NULL`. It seems inconsistent to have special behavior for one key (that's also used in regular typing).

Reply:
@ghislaineguerin I think what @silentninja explained should also be explained better in the spec.

Reply:
@kgodey 

I agree with @silentninja except for _"it would be a text box awaiting a text input rather than a greyed-out box without a cursor"_ (I find that part confusing). 

Here's my take on how a spreadsheet works and also how Mathesar should work:

| current mode | pointer action | pointer target | result |
| -- | -- | -- | -- |
| select | single click | current cell | nothing |
| select | single click | different cell | move selection to new cell |
| select | double click | current cell | switch to edit mode |
| select | double click | different cell | move selection to new cell, switch to edit mode |
| select | triple click | current cell | switch to edit mode, select all |
| select | triple click | different cell | move selection to new cell, switch to edit mode, select all |
| edit | single click | current cell | set cursor position |
| edit | single click | different cell | save changes to current cell, enter select mode, move selection to new cell |
| edit | double click | current cell | select work under cursor |

Now let's add in the `Delete` and `Backspace` behavior

| current mode | DOM `key` | result |
| -- | -- | -- |
| select | any text character | switch to edit mode, set cell contents equal to character entered |
| select | `Backspace` | set cell value to NULL |
| select | `Delete` | set cell value to NULL |
| select | `Shift+Backspace` | nothing |
| select | `Shift+Delete` | nothing |
| edit | `Backspace` | delete character to the left of cursor |
| edit | `Delete` | delete character to the right of cursor |
| edit | `Shift+Backspace` | _discussion pending_ |
| edit | `Shift+Delete` | _discussion pending_ |

Reply:
@seancolsen Thanks for the clear explanation. I find this behavior inconsistent - what if I want to edit the cell and only remove one character? It's confusing that typing `a` in a cell in select mode (for example) adds `a` to the contents of the cell, but typing `Delete` removes the entire contents.

I'd prefer to change the behavior to (only changed table rows are listed):

| current mode | DOM `key` | result |
| -- | -- | -- |
| select | `Backspace` | delete single character |
| select | `Delete` | delete single character |
| edit | `Shift+Backspace` | set cell value to NULL |
| edit | `Shift+Delete` | set cell value to NULL |

Reply:
@kgodey from my understanding if you are in select mode and type `a` you'd be replacing the cell content with the value `a`, not adding to the existing one.

Reply:
@kgodey 

- One clarification:

    > typing `a` in a cell in select mode (for example) adds `a` to the contents of the cell

    No, typing a in `a` cell in select mode _replaces the entire cell contents_ with "a" (and also switches to edit mode). 

- Regarding your suggestion above... that's a bit different from the way that spreadsheets behave. In a spreadsheet, if I'm in select mode and I press `Delete` or `Backspace`, the cell value is set to an empty string.

Reply:
@ghislaineguerin I don't think we should do that either, I think we should err on the side of not entirely replacing data in the database based on quick actions. We care about data integrity. If the user accidentally types `a`, then they can remove the `a` if they make the mistake. If we end up replacing the content of the cell, then they can't recover their previous data, which sounds scary, especially since we don't have undo yet.

Reply:
@seancolsen I didn't see your comment before I responded, but what I said still applies. I don't think we should emulate a spreadsheet in that respect before we have undo functionality.

Reply:
@kgodey good point about the risks given our lack of undo at this stage.

If we don't want `Backspace` and `Delete` to set the cell value to NULL in select mode, then I think we should leave the behavior void at this point so that we can add it in later (once we implement undo) without confusing users who have already been using Mathesar. 

So in that case it would be:

| current mode | DOM `key` | result |
| -- | -- | -- |
| select | `Backspace` | nothing |
| select | `Delete` | nothing |
| select | `Shift+Backspace` | set cell value to NULL |
| select | `Shift+Delete` | set cell value to NULL |

Reply:
I proofread it multiple times and still messed up, edited my comment with the correct answer

Reply:
> No, typing a in a cell in select mode replaces the entire cell contents with "a" (and also switches to edit mode).

I don't think we should replace the cell value, rather the entered character should be appended to the existing value.

Reply:
> the entered character should be appended to the existing value

I don't think there are any spreadsheets that behave that way though.

Reply:
The behaviour I mentioned was the one I was expecting, but Spreadsheets have nailed the UX when it comes to editing, so wouldn't want to ditch that decision and I think it does make sense in a few cases. Do we have a undo shortcut and can it bring the data back?

Reply:
@seancolsen @silentninja I don't think we should perfectly emulate spreadsheets given that we do not have undo yet.

I think that `Backspace` or `Delete` in `select` mode should be similar to entering a text character:

| -- | -- | -- |
| select | any text character | switch to edit mode, append character to text content |
| select | `Backspace` | switch to edit mode, delete single character |
| select | `Delete` |  switch to edit mode, delete single character |

so that wou

Reply:
@ghislaineguerin I find this table view helpful, can we include it in the spec to explain select vs. edit mode?

Reply:
@kgodey

Once we've implemented "undo", would you still want to keep the behavior you're proposing? Or would you want switch to the behavior I'm proposing at that point?

Reply:
@seancolsen I'm not sure, we can decide at that point. It should be after alpha so we'll have user feedback to help make the decision.

Reply:
@kgodey I was asking because your answer would affect my opinion here.

@kgodey Given the consistency in UX across different spreadsheet products, I'm a quite wary of intentionally diverging from those well-established patterns. I understand and agree that we need to implement undo before we can safely implement all of the spreadsheet UX patterns. But I don't want to let the sequence of tasks in our roadmap influence our UX design. If we can't do it "the right way" right now, then I'd rather hold off until we can instead of implementing a UX pattern that will be unfamiliar to users. I can appreciate that data entry will be more cumbersome if I need to switch to edit mode (via double click or `F2`) before editing a cell. But I think that shortcoming is acceptable while we're still in alpha. For alpha I think it's more important to communicate _the direction the product is heading_ than to supply a product that people can rely on for real-world purposes. If we stray from the spreadsheet UX I think our alpha users will receive some subtle communication from the product that direction is not heading exactly towards spreadsheet-like editing, and that will lower peoples' expectations of our stabilized UX.

---

That said, I don't want to harp on this too much. It's a pretty minor detail after all.

If we _do_ go with your proposal, then I'd like to just make the following change 


| current mode | DOM `key` | result |
| -- | -- | -- |
| select | `Delete` | nothing |

Why? Because `Delete` is _forwards_ delete. If we delete the last character in the cell, that would be as if we had placed the cursor one character before the final character, which is weird and arbitrary. If we delete the _first_ character in the cell that seems weird too and inconsistent with any products I can recall using.

We could potentially have `Delete` trigger a "switch to edit mode" action. But for simplicity's sake I'd suggest we leave it unassigned to any action.

Reply:
@seancolsen I see your point but I am unconvinced that the spreadsheet pattern is the "right" way of doing things or that we will end up doing things that way in the end. I think it might be useful for users to understand that we _aren't_ quite a spreadsheet and that we prioritize data integrity. I'd rather take a risk on this new pattern in our alpha release to see if it works well for users. If it doesn't, then we can always switch to the well-established spreadsheet pattern.

So I'd like to go with my proposal. I agree with your amendment, My suggestion was based on `Delete` being the same key as `Backspace` (because that's what it's called on Macs), I agree that forwards `Delete` should do nothing.

Reply:
@kgodey Sounds good :grin:

Reply:
To summarize where we're at now:

| current mode | DOM `key` | result |
| -- | -- | -- |
| select | any text character | switch to edit mode, cursor at end, append character to end of cell contents |
| select | `Backspace` | switch to edit mode, cursor at end, delete last character |
| select | `Delete` | nothing |
| select | `Shift+Backspace` | set cell value to NULL |
| select | `Shift+Delete` | set cell value to NULL |
| edit | `Backspace` | delete character to the left of cursor |
| edit | `Delete` | delete character to the right of cursor |
| edit | `Shift+Backspace` | _discussion pending_ |
| edit | `Shift+Delete` | _discussion pending_ |

Reply:
I'm a little late, but I'm really concerned about a detail here that was assumed, but glossed over:  In spreadsheets, if I enter edit mode, then press `Esc` (without navigating by clicking, tabbing, or returning), the cell returns to its original state.  It seems like we are not planning to have that.  Shouldn't that be the way it works?  We don't actually need "undo" for that, and it's very expected by common users of spreadsheets.  Kind of a freebie "mini-undo".  All we'd need to do is avoid sending any requests until the appropriate navigation has taken place (which we should do anyway).

Given that, I don't see the danger of following the expected "type from select replaces cell value with new character".  All they need to do to "undo" is escape, and users are pretty familiar with that (I tried both Google Sheets and LibreOffice; they each use that setup).

For "delete from select", It's much more efficient to have a LibreOffice sheets style dialog pop up "delete contents?" than to have to delete each character.  This would allow quickly deleting cells without spamming the delete key, but still avoid destroying data without confirmation.  If we want to use a confirmation when "delete" (or backspace) is pressed, it might then make sense to also have a keyboard shortcut for users who know what they're doing to "nullify" a cell without confirmation from select (perhaps shared with edit mode).

Reply:
@mathemancer Please put your comment in https://github.com/centerofci/mathesar/issues/902, we'll fully define keyboard actions for editing then. I think we should go ahead with the proposal here to unblock this spec.

Reply:
@mathemancer the `Esc` has been on my mind too throughout this conversation and I almost brought it up. I would definitely like to implement the `Esc` behavior as part of https://github.com/centerofci/mathesar/issues/902. It certainly mitigates the risk of data loss, but in my opinion it doesn't provide enough mitigation to make me feel comfortable implementing the "type from select replaces cell value with new character" UX pattern in spreadsheets. Why? Because I think we'll also want to implement "save cell contents on blur", meaning if the user clicks away from the cell while in edit mode, we'll save their edits. That makes it pretty easy to mess up.

@kgodey this whole conversation has me elevating my mental prioritization of "undo" quite significantly as it seems to be a crucial factor in implementing the UX patterns which which users are most familiar.

Reply:
Undo has always been high on my mental prioritization for that reason - I don't think users will be able to fully explore and discover features without being able to fix mistakes. But it's also complicated to implement so I think we made the right decision to push it to post-alpha.

Reply:
@kgodey I had a couple more thoughts on this thread last night.

- @ghislaineguerin I'm curious to hear your opinion here. To summarize, we've preliminarily agreed to follow Kriti UX which deviates from standard spreadsheet UX as follows

    | Current mode | DOM `key` | Spreadsheet UX | Kriti's UX |
    | -- | -- | -- | -- |
    | select | any text character | switch to edit mode, **replace** cell contents with character | switch to edit mode, place cursor at end, **append** character to end of cell contents |
    | select | `Backspace` | **delete cell contents** | switch to edit mode, place cursor at end, **delete last character** |

    I'd like us to adhere to the spreadsheet UX, but wait until we have implemented undo before we implement those actions. However, for the sake of moving forward, I've agreed to Kriti's proposal.

    I just want to make sure we're not losing your voice here, Ghislaine since you're the UX person!

- Kriti said:

    > I'd rather take a risk on this new pattern in our alpha release to see if it works well for users. If it doesn't, then we can always switch to the well-established spreadsheet pattern.

    This had me thinking. If we take a risk on a UX pattern that is unfamiliar to users, I think we're going to need to be very aware of this divergence and intentionally seek feedback on it during our user testing. I doubt we'll have the resources to dedicate towards the type of user testing the Microsoft has done with Excel or that Google has done with Sheets. I'm sure those companies have run paid focus groups to observe new users interacting with the product. In my experience building UIs, it's absolutely essential to watch users with the product when seeking feedback on theses subtle UX decisions, but it takes time, especially with a large product. New users will basically never express opinions with the specificity necessary for actionable change on small UX decisions. But _watching_ them use the product will inevitably uncover problems. If we diverge from well established UX patterns that type of focus group testing becomes even more important so we'll need to keep this in mind and consider allocating some budget towards it later this year.

Reply:
>  In my experience building UIs, it's absolutely essential to watch users with the product when seeking feedback on theses subtle UX decisions, but it takes time, especially with a large product. New users will basically never express opinions with the specificity necessary for actionable change on small UX decisions.

It's always been my plan to schedule sessions with users where we watch them use the product after release. I have some experience with running user tests, I read _[Rocket Surgery Made Easy](https://www.amazon.com/Rocket-Surgery-Made-Easy-Yourself/dp/0321657292)_ at my last job, which is a great guide to usability testing and I highly recommend it.

Reply:
@ghislaineguerin and I talked about this on a call, we'll deal with this in the follow up issue #917.